test/test_trace.jl

```diff
@@ -147,9 +153,12 @@ end
     @test findfirst(op -> op isa Loop, tape.ops) !== nothing
 
     _, tape = trace(loop3, 1.0, 3)
-    @test play!(tape, loop3, 2.0, 4) == loop3(2.0, 4)
-    @test compile(tape)(loop3, 2.0, 4) == loop3(2.0, 4)
-    @test findfirst(op -> op isa Loop, tape.ops) !== nothing
+    println("1")
+    println(tape)
+    println(Ghost.to_expr(tape))
+    @show play!(tape, loop3, 2.0, 4) == loop3(2.0, 4)
```

Comment:
This line fails currently on 1.7. It enters an infinite loop: The generated code looks almost the same on 1.6 and 1.7 with only two lines fatally changed at the loop end:
```julia
function f16(x2::Float64, x3::Int64) # julia 1.6
    l9x1 = x3
    l9x2 = x2
    x4 = (tuple)(l9x2)
    while true
        #= /home/jan/.julia/dev/Ghost/src/compile.jl:42 =#
        l9x3 = (>)(l9x1, 1)
        if !l9x3
            #= /home/jan/.julia/dev/Ghost/src/compile.jl:53 =#
            x4 = (tuple)(l9x2)
            break
        end
        l9x4 = (-)(l9x1, 1)
        l10x1 = l9x4
        l10x2 = l9x4
        l9x5 = (tuple)(l10x1, l10x2)
        while true
            #= /home/jan/.julia/dev/Ghost/src/compile.jl:42 =#
            l10x3 = (<)(l10x1, 100)
            if !l10x3
                #= /home/jan/.julia/dev/Ghost/src/compile.jl:53 =#
                l9x5 = (tuple)(l10x1, l10x2)
                break
            end
            l10x4 = (*)(l10x1, l10x2)
            l10x5 = (+)(l10x4, 1)
            l10x1 = l10x5
            l10x2 = l10x2
        end
        l9x6 = (getfield)(l9x5, 1)
        l9x7 = (getfield)(l9x5, 2)
        l9x1 = l9x7
        l9x2 = l9x6
    end
    x5 = (getfield)(x4, 1)
    return x5
end
```
```julia
function f17(x2::Float64, x3::Int64) # julia 1.7
    l9x1 = x3
    l9x2 = x2
    x4 = (tuple)(l9x2)
    while true
        #= /home/jan/.julia/dev/Ghost/src/compile.jl:42 =#
        l9x3 = (>)(l9x1, 1)
        if !l9x3
            #= /home/jan/.julia/dev/Ghost/src/compile.jl:53 =#
            x4 = (tuple)(l9x2)
            break
        end
        l9x4 = (-)(l9x1, 1)
        l10x1 = l9x4
        l10x2 = l9x4
        l9x5 = (tuple)(l10x1, l10x2)
        while true
            #= /home/jan/.julia/dev/Ghost/src/compile.jl:42 =#
            l10x3 = (<)(l10x1, 100)
            if !l10x3
                #= /home/jan/.julia/dev/Ghost/src/compile.jl:53 =#
                l9x5 = (tuple)(l10x1, l10x2)
                break
            end
            l10x4 = (*)(l10x1, l10x2)
            l10x5 = (+)(l10x4, 1)
            l10x1 = l10x5
            l10x2 = l10x2
        end
        l9x6 = (getfield)(l9x5, 1)
        l9x7 = (getfield)(l9x5, 2)
        l9x1 = l9x6 # DIFF 1.6: l9x1 = l9x7
        l9x2 = l9x7 # DIFF 1.6:        l9x2 = l9x6
    end
    x5 = (getfield)(x4, 1)
    return x5
end
```

Reply:
@dfdx any ideas?

Reply:
There should be some difference in `@code_ir f(args...)` that affects loop logic, I'll take a look at it today or tomorrow.

Reply:
Could you please post versions of Julia and all packages? WIth IRTools@v0.4.2 I get an error much earlier, with the latest IRTools@v0.4.3 there are also errors in CUDA-related code, but all tracer tests pass. My setup:

```
Julia 1.7.0-beta2

(Ghost) pkg> st
     Project Ghost v0.1.0
      Status `~/work/Ghost/Project.toml`
  [052768ef] CUDA v3.2.1
  [d360d2e6] ChainRulesCore v0.10.4
  [e30172f5] Documenter v0.27.0
  [7869d1d1] IRTools v0.4.3
  [bac558e1] OrderedCollections v1.4.1
  [37e2e46d] LinearAlgebra
  [8dfed614] Test
  [cf7118a7] UUID
```

Reply:
I can reproduce the problem with this setup:
```
(Ghost) pkg> st
     Project Ghost v0.1.0
      Status `~/.julia/dev/Ghost/Project.toml`
  [052768ef] CUDA v3.3.2
  [d360d2e6] ChainRulesCore v0.10.9
  [e30172f5] Documenter v0.27.3
  [7869d1d1] IRTools v0.4.3
  [bac558e1] OrderedCollections v1.4.1
  [37e2e46d] LinearAlgebra
  [8dfed614] Test
  [cf7118a7] UUIDs

julia> versioninfo()
Julia Version 1.7.0-beta2
Commit b570546b68 (2021-06-20 06:31 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: AMD Ryzen 9 3900 12-Core Processor
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.0 (ORCJIT, znver2)
Environment:
  JULIA_NUM_THREADS = 24
```

Reply:
Weird, even with exactly the same versions I still can't reproduce the error. Could you please post: 

1. Last three commits. Mine are:

```
commit 342b9c560f5f428e59eb00eac098c45816a3854f (HEAD -> julia1.7)
Author: Jan Weidner <jw3126@gmail.com>
Date:   Fri Jul 2 09:44:39 2021 +0200

    wip add julia 1.7 support

commit a6706359f6ddd5f5b0ed8d491e071c5930872dd5 (tag: v0.1.0-docs1)
Author: Andrei Zhabinski <andrei.zhabinski@gmail.com>
Date:   Fri Jun 18 23:22:53 2021 +0300

    Comment out tests on macOS and Windows

commit 0b06435179586638dcb5df0dcee07977ecd290c2
Author: Andrei Zhabinski <andrei.zhabinski@gmail.com>
Date:   Fri Jun 18 02:02:26 2021 +0300

    Fix logo
```

2. Output of `@code_ir loop3(1.0, 3)`.
3. Output of:

```julia
_, tape = trace(loop3, 1.0, 3)
println(tape[V(4)].subtape)
println(tape[V(4)].cont_vars)
```

Reply:
1. I also double-checked that my repo is not dirty
```
commit 342b9c560f5f428e59eb00eac098c45816a3854f (HEAD -> julia1.7, fork/julia1.7)
Author: Jan Weidner <jw3126@gmail.com>
Date:   Fri Jul 2 09:44:39 2021 +0200

    wip add julia 1.7 support

commit a6706359f6ddd5f5b0ed8d491e071c5930872dd5 (tag: v0.1.0-docs1, origin/main, origin/HEAD, main)
Author: Andrei Zhabinski <andrei.zhabinski@gmail.com>
Date:   Fri Jun 18 23:22:53 2021 +0300

    Comment out tests on macOS and Windows

commit 0b06435179586638dcb5df0dcee07977ecd290c2
Author: Andrei Zhabinski <andrei.zhabinski@gmail.com>
Date:   Fri Jun 18 02:02:26 2021 +0300

    Fix logo
```
2.
```
1: (%1, %2, %3)
  br 2 (%3, %2)
2: (%4, %5)
  %6 = %4 > 1
  br 7 unless %6
  br 3
3:
  %7 = %4 - 1
  br 4 (%7)
4: (%8)
  %9 = %8 < 100
  br 6 unless %9
  br 5
5:
  %10 = %8 * %7
  %11 = %10 + 1
  br 4 (%11)
6:
  br 2 (%7, %8)
7:
  return %5
```
3.
```
Tape{Dict{Any, Any}}
  inp %1::Int64
  inp %2::Float64
  %3 = >(%1, 1)::Bool
  %4 = -(%1, 1)::Int64
  %5 = Loop(%4, %4)
  %6 = getfield(%5, 1)::Int64
  %7 = getfield(%5, 2)::Int64

Ghost.Variable[%6, %7]
```

Reply:
Ok, we are getting closer - for me the last test gives me:

```julia
julia> tape[V(4)].cont_vars
2-element Vector{Variable}:
 %7
 %6
```
These are continuation variables - they are passed as inputs to the next iteration, and their order is swapped :facepalm: 

The IR code you posted is identical to mine, but there's one more step affecting it. Could you please run the following:

```julia
ir = IRTools.@code_ir loop3(1.0, 3)
IRTools.expand!(ir)
println(ir)
```
For me it gives:

```julia
1: (%1, %2, %3)
  br 2 (%3, %2)
2: (%4, %5)
  %6 = %4 > 1
  br 7 (%5) unless %6
  br 3 (%4)
3: (%18)
  %7 = %18 - 1
  br 4 (%7, %7)
4: (%8, %17)
  %9 = %8 < 100
  br 6 (%8, %17) unless %9
  br 5 (%8, %17)
5: (%15, %16)
  %10 = %15 * %16
  %11 = %10 + 1
  br 4 (%11, %16)
6: (%13, %14)
  br 2 (%14, %13)       <-- here
7: (%12)
  return %12
```
If my theory is correct, you should see the different order at the line with "<-- here".

Reply:
Seems your theory is correct:
```
1: (%1, %2, %3)
  br 2 (%3, %2)
2: (%4, %5)
  %6 = %4 > 1
  br 7 (%5) unless %6
  br 3 (%4)
3: (%18)
  %7 = %18 - 1
  br 4 (%7, %7)
4: (%8, %17)
  %9 = %8 < 100
  br 6 (%17, %8) unless %9
  br 5 (%17, %8)
5: (%15, %16)
  %10 = %16 * %15
  %11 = %10 + 1
  br 4 (%11, %15)
6: (%13, %14)
  br 2 (%13, %14)    # <-- here
7: (%12)
  return %12
```

Reply:
So the `IRTools.expand!()` is to be blamed. Looking at its code, it's not obvious what exactly goes wrong, but I have one thing to try. Could you please checkout `dfdx/julia1.7-2` branch:

```
git fetch --all
git switch dfdx/julia1.7-2
```

 and repeat the experiment:

```julia
using IRTools

function loop3(a, b)
    while b > 1
        b = b - 1
        a = b
        while a < 100
            a = a * b + 1      
        end
    end
    return a
end

ir = IRTools.@code_ir loop3(1.0, 3)
IRTools.expand!(ir)
println(ir)
```
----
To add more details (mostly for myself when I read it again another day). I'm looking at [this line](https://github.com/FluxML/IRTools.jl/blob/master/src/passes/passes.jl#L150) which iterates variables in `setdiff(uses, defs)` where `uses` is a `Set` and `defs` is a vector. If for whatever reasons (different compilers, different system dependencies, etc.) `setdiff()` creates different order for variables, the order of arguments to the block will also differ from one computer to another. The order of arguments influences the names of branch arguments (which we may confuse with a different order of arguments). In the fix I posted I simply sort the variables in `setdiff()` which should be stable across environments.

Reply:
It seems my_expand! is needs some tweaks. It produces the same code as `IRTools.expand!`. If I run
```julia
using Revise
using IRTools

function loop3(a, b)
    while b > 1
        b = b - 1
        a = b
        while a < 100
            a = a * b + 1
        end
    end
    return a
end

using Ghost
ir = IRTools.@code_ir loop3(1.0, 3)
IRTools.expand!(ir)
println(ir)
Ghost.my_expand!(ir)
@show(VERSION);
```
<details>
  <summary>Full output</summary>

```julia

1: (%1, %2, %3)
  br 2 (%3, %2)
2: (%4, %5)
  %6 = %4 > 1
  br 7 (%5) unless %6
  br 3 (%4)
3: (%18)
  %7 = %18 - 1
  br 4 (%7, %7)
4: (%8, %17)
  %9 = %8 < 100
  br 6 (%8, %17) unless %9
  br 5 (%8, %17)
5: (%15, %16)
  %10 = %15 * %16
  %11 = %10 + 1
  br 4 (%11, %16)
6: (%13, %14)
  br 2 (%14, %13)
7: (%12)
  return %12
VERSION = v"1.6.1"
v"1.6.1"
```

```julia
1: (%1, %2, %3)
  br 2 (%3, %2)
2: (%4, %5)
  %6 = %4 > 1
  br 7 (%5) unless %6
  br 3 (%4)
3: (%18)
  %7 = %18 - 1
  br 4 (%7, %7)
4: (%8, %17)
  %9 = %8 < 100
  br 6 (%17, %8) unless %9
  br 5 (%17, %8)
5: (%15, %16)
  %10 = %16 * %15
  %11 = %10 + 1
  br 4 (%11, %15)
6: (%13, %14)
  br 2 (%13, %14)
7: (%12)
  return %12
VERSION = v"1.7.0-beta2"
v"1.7.0-beta2"

```
</details>
I get the following diff:



![image](https://user-images.githubusercontent.com/7261506/124427693-03822780-dd6c-11eb-8905-304ab5d7c978.png)

Reply:
Oh, wow, I didn't expect it would work 😃 Yes, it would be great if you could create the PR to IRTools. One thing you will need is a test case that is broken by default and works with the fix - saying that it just generates different code doesn't sound like a bug in the first place :) I think the bug should be reproducible if the following also creates an infinite loop:

```julia
using 

function nested_loop(a, b)
    while b > 1
        b = b - 1
        a = b
        while a < 100
            a = a * b + 1      
        end
    end
    return a
end

ir = IRTools.@code_ir loop3(1.0, 3)
IRTools.expand!(ir)
IRTools.evalir(ir, 2.0, 4)
```
If for whatever reason the PR to IRTools isn't merged quickly, we can go with the current solution and leave a `TODO` comment to update it somewhere in the future. 
----
As a side note, IRTools is great, but the community slowly moves to other IR-level tools, mostly concentrated around [JuliaCompilerPlugins](https://github.com/JuliaCompilerPlugins) (slack: #compiler-plugins). Once it matures, I plan to migrate to this set of tools since they are (expected to be) better maintained and bring a faster first trace run. This won't be very soon though :)

Reply:
The generated code of IRTools mwe has the variables switched, but is otherwise equivalent across julia versions for me. So this is maybe not a bug from the perspective of IRTools. Here is what I did:

![image](https://user-images.githubusercontent.com/7261506/124584067-1b34db00-de54-11eb-8dcc-4ae6a81e0fb7.png)

<details>

```julia
# doit.jl
using IRTools
function nested_loop(a, b)
    while b > 1
        b = b - 1
        a = b
        while a < 100
            a = a * b + 1
        end
    end
    return a
end

ir = IRTools.@code_ir nested_loop(1.0, 3)
IRTools.expand!(ir)
@show VERSION
println(ir)
IRTools.evalir(ir, nested_loop, 2.0, 4)
```
```julia
julia> include("doit.jl")
VERSION = v"1.7.0-beta2"
1: (%1, %2, %3)
  br 2 (%3, %2)
2: (%4, %5)
  %6 = %4 > 1
  br 7 (%5) unless %6
  br 3 (%4)
3: (%18)
  %7 = %18 - 1
  br 4 (%7, %7)
4: (%8, %17)
  %9 = %8 < 100
  br 6 (%17, %8) unless %9
  br 5 (%17, %8)
5: (%15, %16)
  %10 = %16 * %15
  %11 = %10 + 1
  br 4 (%11, %15)
6: (%13, %14)
  br 2 (%13, %14)
7: (%12)
  return %12
100
```
```julia
julia> include("doit.jl")
VERSION = v"1.6.1"
1: (%1, %2, %3)
  br 2 (%3, %2)
2: (%4, %5)
  %6 = %4 > 1
  br 7 (%5) unless %6
  br 3 (%4)
3: (%18)
  %7 = %18 - 1
  br 4 (%7, %7)
4: (%8, %17)
  %9 = %8 < 100
  br 6 (%8, %17) unless %9
  br 5 (%8, %17)
5: (%15, %16)
  %10 = %15 * %16
  %11 = %10 + 1
  br 4 (%11, %16)
6: (%13, %14)
  br 2 (%14, %13)
7: (%12)
  return %12
100
```
</details>